Mathlib/LinearAlgebra/Matrix/SchurTriangulation.lean

```diff
@@ -0,0 +1,300 @@
+            by simp ; rfl
+        have hf {bi i' bj j'} (hi : e i = ⟨bi, i'⟩) (hj : e j = ⟨bj, j'⟩) :=
+          calc toMatrixOrthonormal basis f i j
+            _ = toMatrixOrthonormal bE f (e i) (e j) := by
+              rw [f.toMatrixOrthonormal_reindex]
+              rfl
+            _ = ⟪bE (e i), f (bE (e j))⟫_𝕜 := f.toMatrixOrthonormal_apply_apply ..
+            _ = ⟪(B bi i' : E), f (B bj j')⟫_𝕜 := by rw [hB, hB, hi, hj]
+
+        if hj : j < m then
+          let j' : Fin m := ⟨j, hj⟩
+          have hf' {bi i'} (hi : e i = ⟨bi, i'⟩) (h0 : ⟪(B bi i' : E), bV j'⟫_𝕜 = 0) :=
+            calc toMatrixOrthonormal basis f i j
+              _ = ⟪(B bi i' : E), f _⟫_𝕜 := hf hi (Equiv.finAddEquivSigmaCond_true hj)
+              _ = ⟪_, f (bV j')⟫_𝕜 := rfl
+              _ = 0 :=
+                suffices f (bV j') = μ.val • bV j' by rw [this, inner_smul_right, h0, mul_zero]
+                suffices f.HasEigenvector μ (bV j') from this.apply_eq_smul
+                ⟨(bV j').property, fun h => bV.toBasis.ne_zero j' (Subtype.ext h)⟩
+
+          if hi : i < m then
+            let i' : Fin m := ⟨i, hi⟩
+            suffices ⟪(bV i' : E), bV j'⟫_𝕜 = 0 from hf' (Equiv.finAddEquivSigmaCond_true hi) this
+            bV.orthonormal.right (Fin.ne_of_gt hji)
+          else
+            let i' : Fin n := i.subNat' hi
+            suffices ⟪(bW i' : E), bV j'⟫_𝕜 = 0 from hf' (Equiv.finAddEquivSigmaCond_false hi) this
+            V.inner_left_of_mem_orthogonal (bV j').property (bW i').property
+        else
+          have hi (h : i < m) : False := hj (Nat.lt_trans hji h)
+          let i' : Fin n := i.subNat' hi
+          let j' : Fin n := j.subNat' hj
+          calc toMatrixOrthonormal basis f i j
+            _ = ⟪(bW i' : E), f (bW j')⟫_𝕜 :=
+              hf (Equiv.finAddEquivSigmaCond_false hi) (Equiv.finAddEquivSigmaCond_false hj)
+            _ = ⟪bW i', g (bW j')⟫_𝕜 := by simp [g]
+            _ = toMatrixOrthonormal bW g i' j' := (g.toMatrixOrthonormal_apply_apply ..).symm
+            _ = 0 := hg (Nat.sub_lt_sub_right (Nat.le_of_not_lt hj) hji)
+    }
+  else
+    haveI : Subsingleton E := not_nontrivial_iff_subsingleton.mp hE
+    {
+      dim := 0
+      hdim := Module.finrank_zero_of_subsingleton
+      basis := (Basis.empty E).toOrthonormalBasis ⟨nofun, nofun⟩
+      upperTriangular := nofun
+    }
+termination_by Module.finrank 𝕜 E
+decreasing_by exact
+  calc Module.finrank 𝕜 W
+    _ < m + Module.finrank 𝕜 W :=
+      suffices 0 < m from Nat.lt_add_of_pos_left this
+      Submodule.one_le_finrank_iff.mpr μ.property
+    _ = Module.finrank 𝕜 E := hdim
+
+end LinearMap
+
+namespace Matrix
+
+variable {𝕜 : Type*} [RCLike 𝕜] [IsAlgClosed 𝕜]
+/- IMPORTANT: existing `DecidableEq n` should take precedence over `LinearOrder.decidableEq`,
+a.k.a., `instDecidableEq_mathlib`. -/
+variable [Fintype n] [DecidableEq n] [LinearOrder n] (A : Matrix n n 𝕜)
```

Comment:
I'm not sure whether I agree with this; in principle the two should always coincide anyway, and so `[DecidableEq n] [LinearOrder n]` is probably just asking for trouble.